### PR TITLE
gnrc_lorawan: refactor

### PR DESCRIPTION
**Pull request by jia200x**
### Contribution description
This PR:

- Removes the dependency to `gnrc_pktbuf` in GNRC LoRaWAN. Besides reducing the ROM usage a lot, this allows the usage of GNRC LoRaWAN without depending on GNRC specific stuff
- Rename radio events functions to ease maintaining
- Some minor fixes wit "BUSY" logic
- Remove dependency to `gnrc_neterr` (now `gnrc_lorawan` is fully asynchronous)
- Uncrustifies GNRC LoRaWAN files

This step is required in order to add missing class B and class C support.

### Testing procedure
This should be tested with care.
I suggest trying the following:
- Join a LoRaWAN network using ABP and OTAA
- Send some uplinks and downlinks (to port 2). Check that data is received in both directions and that there aren't leaks in the pktbuf (use `gnrc_pktbuf_cmd`module).
- Try sending several packets in a row. It shouldn't crash and the MAC layer should just drop a packet if the MAC is busy.

### Issues/PRs references
~~Depends on #14058~~

### CODE REVIEW

**Comment by leandrolanzieri** [comment]
This needs rebasing

**Comment by jia200x** [comment]
rebased!

**Comment by fjmolinas** [comment]
ping @jia200x!

File: sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mcps.c [hunk 1]
```diff
@@ -54,120 +56,137 @@ uint32_t gnrc_lorawan_fcnt_stol(uint32_t fcnt_down, uint16_t s_fcnt)
...
-        goto out;
+    if(buf < p_mic) {
+        pkt->port = *(buf++);
+        if (buf < p_mic) {
+            pkt->enc_payload.iol_base = buf;
+            pkt->enc_payload.iol_len = p_mic - buf;
+            if(!pkt->port && fopts_length) {
+                DEBUG("gnrc_lorawan: packet with fopts and port == 0. Drop\n");
+                return -1;
+            }
```
[the remark this answers is not in the document]
Comment (jia200x) [reply sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mcps.c]: you are right! I will fix this

[the remark this answers is not in the document]
Comment (jia200x) [reply sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mcps.c]: done!

File: sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c [hunk 2]
```diff
@@ -108,6 +108,30 @@ void gnrc_lorawan_mcps_confirm(gnrc_lorawan_t *mac, mcps_confirm_t *confirm)
...
 
+static void _rx_done(gnrc_lorawan_t *mac)
+{
+    netdev_t *dev = gnrc_lorawan_get_netdev(mac);
+    int bytes_expected = dev->driver->recv(dev, NULL, 0, 0);
+    int nread;
+    struct netdev_radio_rx_info rx_info;
+    gnrc_pktsnip_t *pkt = gnrc_pktbuf_add(NULL, NULL, bytes_expected, GNRC_NETTYPE_UNDEF);
+    if (pkt == NULL) {
+        DEBUG("_recv_ieee802154: cannot allocate pktsnip.\n");
```
[the remark this answers is not in the document]
Comment (jia200x) [reply sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:124]: done

**Comment by jia200x** [comment]
all comments were addressed

**Comment by leandrolanzieri** [comment]
I tested this and it seems that there is a problem when sending multiple packets in a row. After a while the application server stops receiving uplinks. A check on the pktbuf reveals that there are some chunks on it.
```
2020-09-29 13:59:28,355 #  ifconfig
2020-09-29 13:59:28,367 # Iface  3  HWaddr: 26:01:64:36  Frequency: 869524963Hz  BW: 125kHz  SF: 12  CR: 4/5  Link: up 
2020-09-29 13:59:28,376 #            TX-Power: 14dBm  State: SLEEP  Demod margin.: 0  Num gateways.: 0 
2020-09-29 13:59:28,376 #           IQ_INVERT  
2020-09-29 13:59:28,379 #           RX_SINGLE  OTAA  L2-PDU:2303  
2020-09-29 13:59:28,379 #           
send 3 UUUUU 2
2020-09-29 13:59:33,725 #  send 3 UUUUU 2
2020-09-29 13:59:33,726 # Done!
> send 3 UUUUU 2
2020-09-29 13:59:37,554 #  send 3 UUUUU 2
2020-09-29 13:59:37,555 # Done!
> send 3 UUUUU 2
2020-09-29 13:59:40,090 #  send 3 UUUUU 2
2020-09-29 13:59:40,091 # Done!
> send 3 UUUUU 2
2020-09-29 13:59:41,255 #  send 3 UUUUU 2
2020-09-29 13:59:41,256 # Done!
> send 3 UUUUU 2
2020-09-29 13:59:42,174 #  send 3 UUUUU 2
2020-09-29 13:59:42,175 # Done!
> send 3 UUUUU 2
2020-09-29 13:59:42,892 #  send 3 UUUUU 2
2020-09-29 13:59:42,892 # Done!
> send 3 UUUUU 2
2020-09-29 13:59:43,765 #  send 3 UUUUU 2
2020-09-29 13:59:43,766 # Done!
> send 3 UUUUU 2
2020-09-29 13:59:44,528 #  send 3 UUUUU 2
2020-09-29 13:59:44,528 # Done!
send 3 UUUUU 3
2020-09-29 13:59:53,693 #  send 3 UUUUU 3
2020-09-29 13:59:53,694 # Done!
send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:02,083 #  send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:02,083 # Done!
> send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:05,970 #  send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:05,971 # Done!
> send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:07,938 #  send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:07,939 # Done!
> send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:09,206 #  send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:09,207 # Done!
> send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:10,117 #  send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:10,118 # Done!
> send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:16,173 #  send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:16,174 # Done!
> send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:19,165 #  send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:19,166 # Done!
> send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:25,295 #  send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:25,296 # Done!
> send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:28,712 #  send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:28,712 # Done!
> send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:33,518 #  send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:33,519 # Done!
> send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:42,379 #  send 3 UUUUUaaaaaaaaaaaaaaaaaaaaaaaaaaaa 3
2020-09-29 14:00:42,379 # Done!
> send 3 UUU 3
2020-09-29 14:00:45,891 #  send 3 UUUUU 3
2020-09-29 14:00:45,891 # Done!
> pktbuf
2020-09-29 14:00:55,631 #  pktbuf
2020-09-29 14:00:55,639 # packet buffer: first byte: 0x200011c0, last byte: 0x200013c0 (size: 512)
2020-09-29 14:00:55,641 #   position of last byte used: 512
2020-09-29 14:00:55,665 # =========== chunk   0 (0x200011c0 size:  512) ===========
2020-09-29 14:00:55,668 # 00000000  00  00  00  00  80  12  00  20  21  00  00  00  01  00  00  00
2020-09-29 14:00:55,670 # 00000010  00  00  00  00  00  00  00  00  55  55  55  55  55  00  00  00
2020-09-29 14:00:55,688 # 00000020  00  00  00  00  F8  11  00  20  05  00  00  00  01  00  00  00
2020-09-29 14:00:55,690 # 00000030  00  58  6E  84  00  85  D0  8D  55  55  55  55  55  01  00  00
2020-09-29 14:00:55,692 # 00000040  00  00  00  00  18  12  00  20  05  00  00  00  01  00  00  00
2020-09-29 14:00:55,695 # 00000050  00  00  00  00  00  00  00  00  55  55  55  55  55  01  00  00
2020-09-29 14:00:55,712 # 00000060  00  00  00  00  38  12  00  20  05  00  00  00  01  00  00  00
2020-09-29 14:00:55,714 # 00000070  00  00  00  00  00  00  00  00  55  55  55  55  55  01  00  00
2020-09-29 14:00:55,716 # 00000080  00  00  00  00  58  12  00  20  05  00  00  00  01  00  00  00
2020-09-29 14:00:55,734 # 00000090  00  00  00  00  00  00  00  00  55  55  55  55  55  01  00  00
2020-09-29 14:00:55,736 # 000000A0  00  00  00  00  78  12  00  20  05  00  00  00  01  00  00  00
2020-09-29 14:00:55,739 # 000000B0  00  00  00  00  00  00  00  00  55  55  55  55  55  01  00  00
2020-09-29 14:00:55,743 # 000000C0  78  7A  AF  5D  AA  18  00  26  59  4F  81  E3  CE  C8  78  FD
2020-09-29 14:00:55,758 # 000000D0  DF  CD  84  6D  E8  07  79  4A  75  E8  E9  10  3E  83  35  AB
2020-09-29 14:00:55,761 # 000000E0  C0  06  4C  06  00  00  00  00  00  00  00  00  C0  12  00  20
2020-09-29 14:00:55,762 # 000000F0  21  00  00  00  01  00  00  00  00  00  00  00  00  00  00  00
2020-09-29 14:00:55,782 # 00000100  55  55  55  55  55  61  61  61  61  61  61  61  61  61  61  61
2020-09-29 14:00:55,786 # 00000110  61  61  61  61  61  61  61  61  61  61  61  61  61  61  61  61
2020-09-29 14:00:55,788 # 00000120  61  00  00  00  00  00  00  00  00  00  00  00  00  13  00  20
2020-09-29 14:00:55,790 # 00000130  21  00  00  00  01  00  00  00  00  00  00  00  00  00  00  00
2020-09-29 14:00:55,805 # 00000140  55  55  55  55  55  61  61  61  61  61  61  61  61  61  61  61
2020-09-29 14:00:55,807 # 00000150  61  61  61  61  61  61  61  61  61  61  61  61  61  61  61  61
2020-09-29 14:00:55,810 # 00000160  61  00  00  00  00  00  00  00  00  00  00  00  40  13  00  20
2020-09-29 14:00:55,827 # 00000170  21  00  00  00  01  00  00  00  00  00  00  00  00  00  00  00
2020-09-29 14:00:55,831 # 00000180  55  55  55  55  55  61  61  61  61  61  61  61  61  61  61  61
2020-09-29 14:00:55,832 # 00000190  61  61  61  61  61  61  61  61  61  61  61  61  61  61  61  61
2020-09-29 14:00:55,836 # 000001A0  61  00  00  00  00  00  00  00  00  00  00  00  80  13  00  20
2020-09-29 14:00:55,853 # 000001B0  21  00  00  00  01  00  00  00  00  00  00  00  00  00  00  00
2020-09-29 14:00:55,857 # 000001C0  55  55  55  55  55  61  61  61  61  61  61  61  61  61  61  61
2020-09-29 14:00:55,860 # 000001D0  61  61  61  61  61  61  61  61  61  61  61  61  61  61  61  61
2020-09-29 14:00:55,862 # 000001E0  61  00  00  00  00  00  00  00  00  00  00  00  D8  11  00  20
2020-09-29 14:00:55,865 # 000001F0  05  00  00  00  01  00  00  00  00  00  00  00  00  00  00  00
2020-09-29 14:00:55,884 # =========== chunk   1 (0x200011c0 size:  512) ===========
2020-09-29 14:00:55,886 # 00000000  00  00  00  00  80  12  00  20  21  00  00  00  01  00  00  00
2020-09-29 14:00:55,888 # 00000010  00  00  00  00  00  00  00  00  55  55  55  55  55  00  00  00
2020-09-29 14:00:55,905 # 00000020  00  00  00  00  F8  11  00  20  05  00  00  00  01  00  00  00
2020-09-29 14:00:55,908 # 00000030  00  58  6E  84  00  85  D0  8D  55  55  55  55  55  01  00  00
2020-09-29 14:00:55,912 # 00000040  00  00  00  00  18  12  00  20  05  00  00  00  01  00  00  00
2020-09-29 14:00:55,915 # 00000050  00  00  00  00  00  00  00  00  55  55  55  55  55  01  00  00
2020-09-29 14:00:55,931 # 00000060  00  00  00  00  38  12  00  20  05  00  00  00  01  00  00  00
2020-09-29 14:00:55,932 # 00000070  00  00  00  00  00  00  00  00  55  55  55  55  55  01  00  00
2020-09-29 14:00:55,937 # 00000080  00  00  00  00  58  12  00  20  05  00  00  00  01  00  00  00
2020-09-29 14:00:55,953 # 00000090  00  00  00  00  00  00  00  00  55  55  55  55  55  01  00  00
2020-09-29 14:00:55,957 # 000000A0  00  00  00  00  78  12  00  20  05  00  00  00  01  00  00  00
2020-09-29 14:00:55,959 # 000000B0  00  00  00  00  00  00  00  00  55  55  55  55  55  01  00  00
2020-09-29 14:00:55,962 # 000000C0  78  7A  AF  5D  AA  18  00  26  59  4F  81  E3  CE  C8  78  FD
2020-09-29 14:00:55,977 # 000000D0  DF  CD  84  6D  E8  07  79  4A  75  E8  E9  10  3E  83  35  AB
2020-09-29 14:00:55,979 # 000000E0  C0  06  4C  06  00  00  00  00  00  00  00  00  C0  12  00  20
2020-09-29 14:00:55,981 # 000000F0  21  00  00  00  01  00  00  00  00  00  00  00  00  00  00  00
2020-09-29 14:00:56,000 # 00000100  55  55  55  55  55  61  61  61  61  61  61  61  61  61  61  61
2020-09-29 14:00:56,004 # 00000110  61  61  61  61  61  61  61  61  61  61  61  61  61  61  61  61
2020-09-29 14:00:56,007 # 00000120  61  00  00  00  00  00  00  00  00  00  00  00  00  13  00  20
2020-09-29 14:00:56,010 # 00000130  21  00  00  00  01  00  00  00  00  00  00  00  00  00  00  00
2020-09-29 14:00:56,024 # 00000140  55  55  55  55  55  61  61  61  61  61  61  61  61  61  61  61
2020-09-29 14:00:56,027 # 00000150  61  61  61  61  61  61  61  61  61  61  61  61  61  61  61  61
2020-09-29 14:00:56,030 # 00000160  61  00  00  00  00  00  00  00  00  00  00  00  40  13  00  20
2020-09-29 14:00:56,045 # 00000170  21  00  00  00  01  00  00  00  00  00  00  00  00  00  00  00
2020-09-29 14:00:56,048 # 00000180  55  55  55  55  55  61  61  61  61  61  61  61  61  61  61  61
2020-09-29 14:00:56,051 # 00000190  61  61  61  61  61  61  61  61  61  61  61  61  61  61  61  61
2020-09-29 14:00:56,055 # 000001A0  61  00  00  00  00  00  00  00  00  00  00  00  80  13  00  20
2020-09-29 14:00:56,072 # 000001B0  21  00  00  00  01  00  00  00  00  00  00  00  00  00  00  00
2020-09-29 14:00:56,075 # 000001C0  55  55  55  55  55  61  61  61  61  61  61  61  61  61  61  61
2020-09-29 14:00:56,077 # 000001D0  61  61  61  61  61  61  61  61  61  61  61  61  61  61  61  61
2020-09-29 14:00:56,080 # 000001E0  61  00  00  00  00  00  00  00  00  00  00  00  D8  11  00  20
2020-09-29 14:00:56,083 # 000001F0  05  00  00  00  01  00  00  00  00  00  00  00  00  00  00  00
> 2020-09-29 14:01:17,322 # Exiting Pyterm
```

**Comment by jia200x** [comment]
fixed!

**Comment by leandrolanzieri** [comment]
Tested again and is working nicely. I prefer the command waiting for a result of the send, instead of just printing `Done` and silently dropping packets (it even prints the proper error messages now ;) Please go ahead @jia200x and squash!

**Comment by leandrolanzieri** [comment]
As a side note, in the application the port variable is `uint8_t` so it overflows and when a port out of range but greater than 255 is passed to the command it would just wrap and send anyways. But that's not related to these changes, I'd be ok with having this in a follow up PR.

**Comment by jia200x** [comment]
done!

**Comment by leandrolanzieri** [comment]
It looks like AVR does not like `strerror` :/

**Review by leandrolanzieri — APPROVED** [review]
I've tested this on the `lobaro-lorabox`, both joining OTAA and ABP. Uplinks and downlinks with confirmation are working properly. ACK!

**Comment by leandrolanzieri** [comment]
We are currently in soft feature freeze, although this PR introduces no API changes, the rework on the stack is not what I'd call small. 

@bergzand what do you think? do we wait until the hard freeze for this?

**Comment by jia200x** [comment]
ping @bergzand :)

It would be cool to have this (if you think this is in the scope of soft feature freeze). This is a dependency of #14558 (which migrates GNRC LoRaWAN to ztimer)

**Review by bergzand — DISMISSED** [review]
Sorry, blocking this because of the soft freeze, feel free to dismiss after the hard freeze.

**Comment by leandrolanzieri** [comment]
And GO!

**Pull request merged**